integration/kotlinx-coroutines-log4j/src/Log4JThreadContext.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.log4j
+
+import kotlinx.coroutines.*
+import org.apache.logging.log4j.ThreadContext
+import kotlin.coroutines.AbstractCoroutineContextElement
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Context element for [CoroutineContext], enabling the use of Log4J 2's [ThreadContext] with coroutines.
+ *
+ * # Example
+ *
+ * The following example demonstrates usage of this class. All `assert`s pass. Though this example only uses the mapped
+ * diagnostic context, the nested diagnostic context is also supported.
+ *
+ * ```kotlin
+ *  1. runBlocking {
+ *  2.   ThreadContext.put("kotlin", "rocks") // Put a value into the ThreadContext
+ *  3.
+ *  4.   withContext(Log4JThreadContext()) {
+ *  5.     assert(ThreadContext.get("kotlin") == "rocks")
+ *  6.     logger.info(...)   // The ThreadContext contains the mapping here
+ *  7.
+ *  8.     ThreadContext.put("kotlin", "is great")
+ *  9.     launch(Dispatchers.IO) {
+ * 10.       assert(ThreadContext.get("kotlin") == "rocks")
+ * 11.     }
+ * 12.   }
+ * 13. }
+ * ```
+ * It may be surprising that the [ThreadContext] contains the pair (`"kotlin"`, `"rocks"`) at line 10. However, recall
+ * that on line 4, the [CoroutineContext] was updated with the [Log4JThreadContext] element. When, on line 9, a new
+ * [CoroutineContext] is forked from [CoroutineContext] created on line 4, the same [Log4JThreadContext] element from
+ * line 4 is applied. The [ThreadContext] modification made on line 8 is not part of the [state].
+ *
+ * ## Combine with other
+ * You may wish to combine this [ThreadContextElement] with other [CoroutineContext]s.
+ *
+ * ```kotlin
+ * launch(Dispatchers.IO + Log4JThreadContext()) { ... }
+ * ```
+ *
+ * # CloseableThreadContext
+ * [org.apache.logging.log4j.CloseableThreadContext] is useful for automatically cleaning up the [ThreadContext] after a
+ * block of code. The structured concurrency provided by coroutines offers the same functionality.
+ *
+ * In the following example, the modifications to the [ThreadContext] are cleaned up when the coroutine exits.
+ *
+ * ```kotlin
+ * ThreadContext.put("kotlin", "rocks")
+ *
+ * withContext(Log4JThreadContext()) {
+ *   ThreadContext.put("kotlin", "is awesome")
+ * }
+ * assert(ThreadContext.get("kotlin") == "rocks")
+ * ```
```

Comment:
The above is a misleading and fragile example. In case of any suspension in the outer code or inside of `withContext` the behavior of this code will change, as the thread context will get restore to the state that is captured in the coroutine. 

This is true for any pattern like this:
```
ThreadContext.put("kotlin", "rocks")
// <-- adding suspension here breaks it
withContext(Log4JThreadContext()) { ... }
```

What I suggest is to design a convenient API to specify additional parameters right in the `Log4JThreadContext`, so that you can write the above code in a less-fragile way:
```
withContext(Log4JThreadContext(mapOf("kotlin" to "rockts"))) { ... }
```
(or something similar)